spring-vaadin-security/src/main/java/org/vaadin/spring/security/provider/AbstractAnnotationAccessDelegate.java

```diff
@@ -0,0 +1,166 @@
+import org.springframework.security.authentication.InsufficientAuthenticationException;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.util.MethodInvocationUtils;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.vaadin.spring.navigator.ViewProviderAccessDelegate;
+import org.vaadin.spring.security.VaadinSecurity;
+import org.vaadin.spring.security.VaadinSecurityAware;
+
+import com.vaadin.navigator.View;
+import com.vaadin.ui.UI;
+
+/**
+ * Abstract implementation to provide easy handling of annotation secured objects
+ * 
+ * @author Marko Radinovic (markoradinovic79@gmail.com)
+ * @author Gert-Jan Timmer (gjr.timmer@gmail.com)
+ * <br><br>
+ * Initial code:<a href="https://github.com/markoradinovic/Vaadin4Spring-MVP-Sample-SpringSecurity">https://github.com/markoradinovic/Vaadin4Spring-MVP-Sample-SpringSecurity</a>
+ */
+public abstract class AbstractAnnotationAccessDelegate implements ApplicationContextAware, VaadinSecurityAware, InitializingBean, ViewProviderAccessDelegate {
+
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+	
+	protected ApplicationContext applicationContext;
+	protected VaadinSecurity security;
+	
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;		
+	}
+	
+	@Override
+	public void setVaadinSecurity(VaadinSecurity vaadinSecurity) {
+		this.security = vaadinSecurity;
+	}
+	
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		logger.trace("> afterPropertiesSet()");
+		
+		Assert.notNull(applicationContext);
+		Assert.notNull(security);
+		
+		logger.trace("< afterPropertiesSet()");
+	}
+	
+	/**
+	 * Helper class which decides if a object is annotated and configures the access decision manager
+	 * with the correct values.
+	 * 
+	 * @param beanName {@link String} bean name
+	 * @param ui {@link UI} UI instance
+	 * @param annotation Annotation class which to check
+	 * @param annotationAttribute Annotation attribute which contains the arguments for the access decision manager
+	 * 
+	 * @return <code>true</code> when access is granted by the access decision manager, <code>false</code> when access is denied by the access
+	 * decision manager.
+	 */
+	protected boolean isAccessGrantedForAnnotation(String beanName, UI ui, Class<? extends Annotation> annotation, String annotationAttribute) {
+		
+		Annotation annotationClass = applicationContext.findAnnotationOnBean(beanName, annotation);
+		
+		if ( annotationClass == null ) {
+        	logger.trace("{} annotation not present on view", annotation.getSimpleName());
+            return true;
+        } else if ( security.hasAccessDecisionManager() ) {
+
+        	logger.trace("DecisionManager present");
+        	
+            final Class<?> targetClass = AopUtils.getTargetClass(applicationContext.getBean(beanName));
```

Comment:
If the view happens to be prototype scoped, a new instance will be created each time you ask for it from the application context. If the view happens to be view scoped, you will get an exception since the view scope is not active during the first security check.